Sidebar_Chart_Suggestions.gs

```diff
@@ -0,0 +1,252 @@
+      if(typeof(data[row_idx][j])==='string'){
+          count++;   
+      }
+      Logger.log(count);
+      if(count>=10 || count==row_length){ 
+          return true;
+      }
+  }  
+  return false;
+}
+
+function isStringCol(data, col_idx){
+  Logger.log('string col '+ col_idx);
+  var count = 0;
+  var col_length = data.length;
+  for (var i = 0; i <col_length; i++){
+    if(typeof(data[i])==='undefined' || typeof(data[i][col_idx])==='undefined'){
+         return false;
+    }
+    if(typeof(data[i][col_idx])==='string'){
+          count++;   
+     }
+    if(count>=10 || count==col_length){
+      return true;
+    }
+  }
+  return false;
+}
+
+function isValidDate(date) {
+  return date && Object.prototype.toString.call(date) === "[object Date]" && !isNaN(date);
+}
+
+function isTimeDateRow(data){
+   var count = 0;
+   var row_length = data[0].length;
+   for(var j=0; j<row_length; j++){ 
+//      if(typeof(data[0][j])==='undefined'){
+//         return false;
+//      }
+      if(isValidDate(data[0][j])){
+          count++;   
+      }
+      if(count>=10 || count==row_length){
+          return true;
+      }
+   }  
+   return false;
+}
+
+function isTimeDateCol(data){
+  var count = 0;
+  var col_length = data.length;
+  for (var i = 0; i <col_length; i++){
+//    if(typeof(data[i])==='undefined'){
+//         return false;
+//    }
+    if(isValidDate(data[i][0])){
+          count++;   
+    }
+    if(count>=10 || count==col_length){
+          return true;
+    }
+  }
+  return false;
+}
+
+function isPieCol(data){
+  var sum=0;
+  var lastCidx = data.length-1; //last elt index in a column
+  var no_of_cols = data[0].length;
+  var no_of_rows = data.length;
+  if( no_of_cols == 2 && (isStringCol(data,0) && isNumberCol(data,1) || isStringCol(data,0) && isNumberCol(data,1))){
+    return 1;
+  }
+  for (var i = 0; i <data.length-1; i++){
+    var ratio = (Number(data[i][0])/Number(data[lastCidx][0]));
+    if(ratio>=0 && ratio<=1){
+      sum += Number(data[i][0]);
+    }
+    else{
+      return 0;
+    }
+  }
+  if(sum==data[lastCidx][0]){
+    return 2;
+  }
+  return 0;
+}
+
+function isPieRow(data){
+  Logger.log('pie2');
+  var sum=0;
+  var lastRidx = data[0].length-1; //last elt index in a row
+  var no_of_cols = data[0].length;
+  var no_of_rows = data.length;
+  if( no_of_rows == 2 && isStringRow(data,0) && isNumberRow(data,1)){
+    return true;
+  }
+  for (var i = 0; i <lastRidx; i++){
+    var ratio = (Number(data[0][i])/Number(data[0][lastRidx]));
+    if(ratio>=0 && ratio<=1){
+      sum += Number(data[0][i]);
+    }
+    else{
+      return false;
+    }
+  }
+  if(sum==data[0][lastRidx]){
+    return true;
+  }
+  return false;
+}
+
+function GiveSuggestion(range){
+  Logger.log('suggestion fn');
+  Logger.log(range);
+   var sheet = SpreadsheetApp.getActiveSheet();
+  var searchRange = sheet.getRange(range); 
+  var data = searchRange.getValues();
+  var suggest = [];
+  
+  //traverse first row of range
```

Comment:
// Traverse

space between // and first character and start with a capital letter. Also in other places